R/type_factories.R

```diff
@@ -0,0 +1,155 @@
+#' @export
+type_loess = function(x, y, ...) {
+    fun = function(x, y, ...) {
+        if (missing(x) || missing(y)) return("l")
+        dat = data.frame(x, y)
+        fit = stats::loess(y ~ x, data = dat)
+        nd = data.frame(x = seq(min(x), max(x), length.out = 1000))
+        y = stats::predict(fit, newdata = nd)
+        out = list(y = y, x = nd$x)
+        return(out)
+    }
+    return(fun)
+}
+
+
+#' Type factory: GLM
+#'
+#' Call this function in the `type` argument of `tinyplot()` to display a 
+#' GLM-fitted version of the relationship between `y` and `x`.
+#'
+#' @inheritParams tinyplot
+#' @inheritParams type_spline
+#' @inheritParams stats::glm
+#' @details
+#' First, we fit a model using the [stats::glm]. The predicted values of `y` are then obtained by calling [marginaleffects::predictions].
+#'
+#' @examples
+#' plt(vs ~ mpg, type = type_glm(family = binomial), data = mtcars)
+#' @export
+type_glm = function(x, y, family = stats::gaussian(), conf_level = 0.95) {
+    se <- if (isFALSE(conf_level)) FALSE else TRUE
+    conf_level <- if (isFALSE(conf_level)) 0.95 else conf_level
+    assert_dependency("marginaleffects")
+    fun = function(x, y, ...) {
+        if (missing(x) || missing(y)) {
+            if (isTRUE(se)) {
+                return("ribbon")
+            } else {
+                return("l")
+            }
+        }
+        dat = data.frame(x, y)
+        fit = stats::glm(y ~ x, family = family, data = dat)
+        nd = data.frame(x = seq(min(x), max(x), length.out = 1000))
+        p = marginaleffects::predictions(fit, newdata = nd, vcov = se, conf_level = conf_level)
+        out = list(x = p$x,
+            y = p$estimate,
+            ymin = p$conf.low,
+            ymax = p$conf.high)
+        return(out)
+    }
+    return(fun)
+}
+
+
+#' Type factory: LM
+#'
+#' Call this function in the `type` argument of `tinyplot()` to display a 
+#' LM-fitted version of the relationship between `y` and `x`.
+#'
+#' @inheritParams tinyplot
+#' @inheritParams type_spline
+#' @details
+#' First, we fit a model using the [stats::lm]. The predicted values of `y` are then obtained by calling [marginaleffects::predictions].
+#'
+#' @examples
+#' plt(hp ~ mpg, type = type_lm(), data = mtcars)
+#' @export
+type_lm = function(x, y, conf_level = 0.95) {
+    se <- if (isFALSE(conf_level)) FALSE else TRUE
+    conf_level <- if (isFALSE(conf_level)) 0.95 else conf_level
+    assert_dependency("marginaleffects")
+    fun = function(x, y, ...) {
+        if (missing(x) || missing(y)) {
+            if (isTRUE(se)) {
+                return("ribbon")
+            } else {
+                return("l")
+            }
+        }
+        dat = data.frame(x, y)
+        fit = stats::lm(y ~ x, data = dat)
+        nd = data.frame(x = seq(min(x), max(x), length.out = 1000))
+        p = marginaleffects::predictions(fit, newdata = nd, vcov = se, conf_level = conf_level)
```

Comment:
Again, we should be able to stick to pure base R with something like `p = predict(fit, interval = "confidence")`.